x/ccv/provider/keeper/relay.go

```diff
@@ -126,80 +127,86 @@ func (k Keeper) EndBlockVSU(ctx sdk.Context) {
-				err := utils.SendIBCPacket(
-					ctx,
-					k.scopedKeeper,
-					k.channelKeeper,
-					channelID,          // source channel id
-					ccv.ProviderPortID, // source port id
-					packetData.GetBytes(),
-					k.GetParams(ctx).CcvTimeoutPeriod,
-				)
-				if err != nil {
-					panic(fmt.Errorf("packet could not be sent over IBC: %w", err))
-				}
-				// set the VSC send timestamp for this packet
-				k.SetVscSendTimestamp(ctx, chainID, packetData.ValsetUpdateId, ctx.BlockTime())
-			} else {
-				// store the packet data to be sent once the CCV channel is established
-				k.AppendPendingVSC(ctx, chainID, packetData)
-			}
+			k.SendVSCPacketsToChain(ctx, chainID, channelID)
 		}
-		return true // do not stop the iteration
+		return true // continue iterating chains
 	})
-	k.IncrementValidatorSetUpdateId(ctx)
 }
 
-// Sends all pending ValidatorSetChangePackets to the specified chain
-func (k Keeper) SendPendingVSCPackets(ctx sdk.Context, chainID, channelID string) {
-	pendingPackets := k.ConsumePendingVSCs(ctx, chainID)
+// SendVSCPacketsToChain sends all queued ValidatorSetChangePackets to the specified chain
+func (k Keeper) SendVSCPacketsToChain(ctx sdk.Context, chainID, channelID string) {
+	pendingPackets := k.GetPendingVSCs(ctx, chainID)
 	for _, data := range pendingPackets {
-		// send packet over IBC
-		err := utils.SendIBCPacket(
+		// prepare to send the data to the consumer
+		packet, channelCap, err := utils.PrepareIBCPacketSend(
 			ctx,
 			k.scopedKeeper,
 			k.channelKeeper,
 			channelID,          // source channel id
 			ccv.ProviderPortID, // source port id
 			data.GetBytes(),
-			k.GetParams(ctx).CcvTimeoutPeriod,
+			k.GetCCVTimeoutPeriod(ctx),
 		)
 		if err != nil {
+			// something went wrong when preparing the packet
+			panic(fmt.Errorf("packet could not be prepared for IBC send: %w", err))
+		}
+
+		err = k.channelKeeper.SendPacket(ctx, channelCap, packet)
+
+		if err != nil && clienttypes.ErrClientNotActive.Is(err) {
+			// IBC client is expired!
+			// leave the packet data stored to be sent once the client is upgraded
+			// the client cannot expire during iteration (in the middle of a block)
+			return
+		} else if err != nil {
 			panic(fmt.Errorf("packet could not be sent over IBC: %w", err))
+
 		}
 		// set the VSC send timestamp for this packet;
 		// note that the VSC send timestamp are set when the packets
 		// are actually sent over IBC
 		k.SetVscSendTimestamp(ctx, chainID, data.ValsetUpdateId, ctx.BlockTime())
 	}
+	k.DeletePendingVSCs(ctx, chainID)
+}
+
+// queueVSCPackets queues latest validator updates for every registered consumer chain
+func (k Keeper) QueueValidatorUpdates(ctx sdk.Context) {
+	valUpdateID := k.GetValidatorSetUpdateId(ctx) // curent valset update ID
+	valUpdates := k.stakingKeeper.GetValidatorUpdates(ctx)
+
+	k.IterateConsumerChains(ctx, func(ctx sdk.Context, chainID, clientID string) (stop bool) {
+		// check whether there are changes in the validator set;
```

Comment:
This comment should be above `	valUpdates := k.stakingKeeper.GetValidatorUpdates(ctx)`

Reply:
Done